### PR TITLE
[SPARK-16059][R] Add `monotonically_increasing_id` function in SparkR

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR adds `monotonically_increasing_id` column function in SparkR for API parity.
After this PR, SparkR supports the followings.

``` r
> df <- read.json("examples/src/main/resources/people.json")
> collect(select(df, monotonically_increasing_id(), df$name, df$age))
  monotonically_increasing_id()    name age
1                             0 Michael  NA
2                             1    Andy  30
3                             2  Justin  19
```
## How was this patch tested?

Pass the Jenkins tests (with added testcase).

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @shivaram , @felixcheung , @sun-rui .
Could you review this PR?

File: R/pkg/R/functions.R [hunk 1]
```diff
@@ -911,6 +911,33 @@ setMethod("minute",
...
+#' This is equivalent to the MONOTONICALLY_INCREASING_ID function in SQL.
+#'
+#' @rdname monotonically_increasing_id
+#' @name monotonically_increasing_id
+#' @family misc_funcs
+#' @export
+#' @examples \dontrun{select(df, monotonically_increasing_id())}
+setMethod("monotonically_increasing_id",
+          signature(x = "missing"),
+          function(x) {
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: `function()` instead

Comment (dongjoon-hyun) [inline R/pkg/R/functions.R]: Oops. Sure!

File: R/pkg/R/functions.R [hunk 2]
```diff
@@ -911,6 +911,33 @@ setMethod("minute",
             column(jc)
           })
 
+#' monotonically_increasing_id
+#'
+#' Return a column that generates monotonically increasing 64-bit integers.
+#'
+#' The generated ID is guaranteed to be monotonically increasing and unique, but not consecutive.
+#' The current implementation puts the partition ID in the upper 31 bits, and the record number
+#' within each partition in the lower 33 bits. The assumption is that the data frame has
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: does it matter if this is "data frame" vs "DataFrame" or "SparkDataFrame"?

Comment (dongjoon-hyun) [inline R/pkg/R/functions.R]: Nice catch! We should use `SparkDataFrame` here for consistency.
I'll fix this, too!

**Comment by felixcheung** [comment]
LGTM

**Comment by shivaram** [comment]
Thanks @dongjoon-hyun and @felixcheung -- Merging this to master and branch-2.0

**Comment by dongjoon-hyun** [comment]
Thank you, @shivaram and @felixcheung !

**Pull request closed without merging**